### DISCUSSION TITLE
字段模型的unmount属性是否需要决定字段跳过校验

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#2372] [Show and tell]
之前没有想清楚unmount与字段校验的关系，接收了该PR。https://github.com/alibaba/formily/pull/2066

导致引起ArrayTable无法分页的问题：https://github.com/alibaba/formily/discussions/2358

目前需要回滚该逻辑，主要是因为 **组件的卸载，其实并不代表字段被删除**，仅仅只是试图上的一个状态体现，如果基于该状态来跳过校验，那必然会存在很多场景上的冲突问题，最典型的就是ArrayTable分页，或者虚拟滚动场景。

目前唯一确定的逻辑是，字段的display为none或者hidden时，和pattern不为editable时，是肯定会跳过校验的，因为这个是100%可以确定当前字段是不需要校验的

需要注意的是，回滚该逻辑，会造成Break Change，大家需要关注，自己是否存在依赖受控卸载字段来控制字段校验的逻辑，如果存在，请改为正常的修改字段的display:none/hidden这种模式来控制

涉及到的版本为2.0.0-rc.5以后的版本

**Comment by liuweiGL** [comment]
我觉得这个 PR 不能撤回吧：
1. 字段联动的场景下把 UI 隐藏了，这个时候继续校验隐藏掉的字段是否合理？
2. 分页 table 无法校验未展示出来的数据，我感觉这是个伪命题；数据都没展示出来，就算校验了也没啥意义吧，用户都看不到错误信息。我认为这种需求应该是项目层面自己封装一下，在触发分页之前确保当前的数据满足校验。

**Reply by sage-wen** [reply]
我和你的看法是一致的，数据隐藏，就应该在数据隐藏前处理好校验逻辑，否则数据隐藏后数据不合法，无法定位到不合法数据，只能类似弹框提醒，这样对用户是不友好的

**Reply by janryWang** [reply]
字段模型的删除与组件卸载其实并不是一个幂等关联关系，所以组件卸载的触发行为是否要控制校验，也不是一个绝对完备的行为，比如给组件加个随机key，每次重新渲染的时候字段就会频繁触发unmount，如果以unmount为判断标准，那就很容易出现异常问题，

**Reply by MisicDemone** [reply]
应该通过field的state去判断是否校验，而不是根据UI是否隐藏。不然node下想单纯做个校验服务还得手动控制mount/unmount了。

**Reply by lyx9439** [reply]
校验校验的是数据，不是视图。哪些数据应该被校验，应该由状态来控制：一般来讲和显隐状态（指「应该被显示」，而非DOM渲染。即：被虚拟滚动等优化掉的DOM是应该被显示、应该被校验的）一致即可。至于组件卸载时是否需要校验，应该由业务在卸载时更改显隐状态（如果实在需要的话，但这种一般会导致错误），通过状态来控制哪些数据需要被校验。

**Reply by lyx9439** [reply]
不管是分页，还是虚拟滚动，都只是UI的展现形式而已，实际上提交的时候，只要是「理论上可见」（想象一下没有分页和虚拟滚动，把所有字段都平铺在一个屏幕里）的，都应该被校验

**Comment by liuweiGL** [comment]
或者 `createForm` 提供一个参数来控制 unmount 状态是否继续校验。

**Comment by janryWang** [comment]
这个问题只有在用户受控setState去控制某个Field组件显示隐藏会导致提交保留校验，但是如果用到了这种模式，用户自己手动删除字段，其实才是合理的，参考 https://v2.formilyjs.org/zh-CN/guide/advanced/controlled#%E5%AD%97%E6%AE%B5%E7%BA%A7%E5%8F%97%E6%8E%A7
